### PR TITLE
[Structural] Defining initial strains in beams

### PR DESCRIPTION
**Pull request by clazaro**
**📝 Description**
This PR is a proposal on how to  define initial unit elongations and  unit rotations in beam elements.

The constitutive behaviour of Kratos beam elements is defined in custom functions inside the element code - e.g. `CalculateElementForces()` in the 3D2N element.

My proposal is to define the initial strains by means of three element variables for the 3D case (two for 2D):
`BEAM_INITIAL_UNIT_ELONGATION`
`BEAM_INITIAL_UNIT_ROTATION_2`
`BEAM_INITIAL_UNIT_ROTATION_3`

The code in the PR shows just the implementation of the first one: I registered the new variable and modified the constitutive equation for the axial force in `CalculateElementForces()`, that now gets the initial term as follows (compare with Krenk, eq. 5.56):
image

You can test the code with the attached minimal example:
[Beam_initial_strains_MWE.zip](https://github.com/KratosMultiphysics/Kratos/files/7671264/Beam_initial_strains_MWE.zip)

Intial unit rotations can be implemented similarly by subtracting the unit rotation times the lenght in the corresponding symmetric moment equation (see again Krenk, eq. 5.56). 

If this approach makes sense, I would modify the cpp code, add tests and a process to define the variables.

### CODE REVIEW

**Review by loumalouomega — CHANGES_REQUESTED** [review]
This is fine, but I will transform that minimal example in a test (how big it is?)

**Comment by loumalouomega** [comment]
> This is fine, but I will transform that minimal example in a test (how big it is?)

I just check and it is  elements, can be added as a test no problem

**Comment by clazaro** [comment]
... resuming this PR ...
I have slightly changed the approach to introduce initial strains in the cr beam 3D2N:
I have created and registered a single vector variable `BEAM_INITIAL_STRAIN_VECTOR` that collects all initial strains. With this, only one `if` statement is needed to check for initial strains in the element code.
The other decision has been to limit the possible initial strains to three (unit elongation and unit rotations ('curvatures') wrt axes 2 and 3) in the 3D case, with which thermal actions can be straightforwardly modelled. This may be controversial because there are six generalized strains in the 3D beam, but I don't think that an initial unit twisting rotation or initial shearing strains have much physical sense.
The equations supporting this change are (again, see  Krenk, eq. 5.56):
equations
where u is the axial displacement, theta_s are the (symmetric) flexural rotations wrt axes 2 and 3, and chi_0 are the corresponding unit rotations ('curvatures')

**Comment by clazaro** [comment]
I have also included a unit test with 2 beam elements. The left end is fully fixed and the right one is free. Initial strains are defined in the mdpa, and displacements of the free end are checked.
One thing that I wasn't expecting is that +5 steps are needed to achieve proper value of the internal forces and reactions (zero in this case).
You can check it with the attached files (same as the unit test plus the `MainKratos.py` printing the results to screen):
[Beam3D2N_initial_strain_test.zip](https://github.com/KratosMultiphysics/Kratos/files/7960603/Beam3D2N_initial_strain_test.zip)

**Comment by clazaro** [comment]
If this approach makes sense, I will modify the linear element and the 2D versions.
In addition, a specific process could be also prepared to input the strains. It must be different from `set_initial_state_process` though, because the beam is not using a proper CL.

**Comment by clazaro** [comment]
@loumalouomega, I don't know why the checks are in waiting status. Any idea on how to make them run?

**Comment by loumalouomega** [comment]
> @loumalouomega, I don't know why the checks are in waiting status. Any idea on how to make them run?

I did some commit to trigger, should work now

**Comment by clazaro** [comment]
> This is fine, but I will transform that minimal example in a test (how big it is?)

(I am resuming (again) this PR.) I alraady included (some time ago) the requested test and it works well. I guess that before merging, I have to do equivalent changes in the remaining beam elements.

**Comment by clazaro** [comment]
@KlausBSautter The last commits include the modification of the beam2D2N and the corresponding test. The test structures are cantilevers with 2 elements subjected to initial axial strain and initial "curvature". 
The test for the beam3D3N works fine. However, in the test for the 2D element, the displacements are fine, but the section forces are not zero (as they should be). I attach zip files with each test including `MainKratos.py` with instructions to print displacements and section forces.

[crBeam3D2N_initial_strain_test.zip](https://github.com/KratosMultiphysics/Kratos/files/10912211/crBeam3D2N_initial_strain_test.zip)
[crBeam2D2N_initial_strain_test.zip](https://github.com/KratosMultiphysics/Kratos/files/10912215/crBeam2D2N_initial_strain_test.zip)

It would be great if you can have a look!

**Comment by KlausBSautter** [comment]
> ... resuming this PR ... I have slightly changed the approach to introduce initial strains in the cr beam 3D2N: I have created and registered a single vector variable `BEAM_INITIAL_STRAIN_VECTOR` that collects all initial strains. With this, only one `if` statement is needed to check for initial strains in the element code. The other decision has been to limit the possible initial strains to three (unit elongation and unit rotations ('curvatures') wrt axes 2 and 3) in the 3D case, with which thermal actions can be straightforwardly modelled. This may be controversial because there are six generalized strains in the 3D beam, but I don't think that an initial unit twisting rotation or initial shearing strains have much physical sense. The equations supporting this change are (again, see Krenk, eq. 5.56): equations where u is the axial displacement, theta_s are the (symmetric) flexural rotations wrt axes 2 and 3, and chi_0 are the corresponding unit rotations ('curvatures')

I like this approach. If someone in the future needs the other initial strains, they can be easily added.

**Comment by KlausBSautter** [comment]
> @KlausBSautter The last commits include the modification of the beam2D2N and the corresponding test. The test structures are cantilevers with 2 elements subjected to initial axial strain and initial "curvature". The test for the beam3D3N works fine. However, in the test for the 2D element, the displacements are fine, but the section forces are not zero (as they should be). I attach zip files with each test including `MainKratos.py` with instructions to print displacements and section forces.
> 
> [crBeam3D2N_initial_strain_test.zip](https://github.com/KratosMultiphysics/Kratos/files/10912211/crBeam3D2N_initial_strain_test.zip) [crBeam2D2N_initial_strain_test.zip](https://github.com/KratosMultiphysics/Kratos/files/10912215/crBeam2D2N_initial_strain_test.zip)
> 
> It would be great if you can have a look!

I think we already discussed this, and could not figure out the problem. I will check the changed files in general and approve the PR provisionally

File: applications/StructuralMechanicsApplication/tests/InitialStateElasticity/initial_strain_cr_beam3D2N_test_parameters.json
Comment (KlausBSautter) [inline applications/StructuralMechanicsApplication/tests/InitialStateElasticity/initial_strain_cr_beam3D2N_test_parameters.json]: Did you check this with the `and_criterion`?

**Review by KlausBSautter — DISMISSED** [review]

**Comment by KlausBSautter** [comment]
Looks nice! Thanks for further developing the beam. I see the tests are also non-linear now, did this fix the problems you experienced?

**Comment by AlejandroCornejo** [comment]
One comment, Is not possible to storee these initial values inside the `initial_state`  class embedded in the base `constitutive_law`?

In solid elements this is the way that we designed and fully available in the core. 

for example inside the constitutive_law.h:

~~~cpp
    /**
     * @brief Adds the initial stress vector if it is defined in the InitialState
     */
    template<typename TVectorType>
    void AddInitialStressVectorContribution(TVectorType& rStressVector)
    {
        if (this->HasInitialState()) {
            const auto& r_initial_state = GetInitialState();
            noalias(rStressVector) += r_initial_state.GetInitialStressVector();
        }
    }
~~~

**Comment by clazaro** [comment]
@AlejandroCornejo, I am not sure if it is possible: as of now, crBeams don't use `constitutive_law`. The constitutive behaviour is implemented within the element.

**Comment by AlejandroCornejo** [comment]
> @AlejandroCornejo, I am not sure if it is possible: as of now, crBeams don't use `constitutive_law`. The constitutive behaviour is implemented within the element.

Fair enough :)

File: applications/StructuralMechanicsApplication/tests/InitialStateElasticity/initial_strain_cr_beam3D2N_test_parameters.json
Comment (clazaro) [reply applications/StructuralMechanicsApplication/tests/InitialStateElasticity/initial_strain_cr_beam3D2N_test_parameters.json]: Yes. It works, with proper tolerances. I have just pushed a modified test with the `and_criterion`

**Comment by clazaro** [comment]
@KlausBSautter Many thanks for your review and comments. I have also modified the test for the 2D beam to have only one time step. It converges now: forces and reactions are still different from zero but very small (on the order of 1e-4 or 1e-5). Increasing the number of steps, not only does not solve the problem, it renders worse results (previous version). I guess the reason is not in the beam formulation.

**Comment by clazaro** [comment]
@loumalouomega, I added tests for the 2D and 3D beam as requested. Hope it is fine now!

**Comment by loumalouomega** [comment]
> @loumalouomega, I added tests for the 2D and 3D beam as requested. Hope it is fine now!

Okay from my side then

**Comment by clazaro** [comment]
Thanks @loumalouomega. I think that your former request is blocking merging...
Once unblocked, I guess I can merge.

**Review by loumalouomega — APPROVED** [review]
I forgot to approve, jeje

**Comment by clazaro** [comment]
For the record: initial strains can be defined for the non-linear 2D and 3D crBeams. The feature needs yet to be implemented for the linear elements.

**Pull request merged**
